File: src/lib/mag_gal.cpp
/*
    10/12/2014
    Construct the different models of evolution used to measure the photo-z

    2 options : mag_gal
                        -c zphot.para    : config file
                        -t (G/g or Q/q)  : GAL/QSO library
*/

#include <getopt.h>  // get option line commands getopt_long

#include <cmath>     // for the log
#include <cstring>   // Use string in the old c format
#include <ctime>     // date
#include <fstream>   // print output file
#include <iomanip>   // std::set precision
#include <iostream>  // print standard file
#include <sstream>
#include <string>  // use string
#include <vector>  // manipulate vector

// Le Phare
#include "SED.h"
#include "globals.h"  // global variables
#include "keyword.h"  // to read the keyword in the command line/para file
#include "mag.h"      // SED class

using namespace std;

// START
int main(int argc, char *argv[]) {
  int num_threads;

#ifdef _OPENMP
#pragma omp parallel
  num_threads = omp_get_num_threads();
  cout << "Running with OpenMP on " << num_threads << " threads!" << endl;
#endif

  // to estimate the times needed to run the code
  // since clock() returns the sum of all threads, we need to use the openmp
  // timer to measure the time
#ifdef _OPENMP
  double start = omp_get_wtime();
#else
  clock_t t1 = clock();
#endif

  /*
  ENVIRONMENT VARIABLES LEPHAREDIR and LEPHAREWORK
  */
  get_lephare_env();

  /*
  ANALYSE KEYWORDS
  */

  // List of the keywords to be found in the config file/command line
  string list_keywords[] = {"t",           "COSMOLOGY",     "FILTER_FILE",
                            "MAGTYPE",     "EXTINC_LAW",    "EB_V",
                            "MOD_EXTINC",  "Z_STEP",        "GAL_LIB_IN",
                            "QSO_LIB_IN",  "STAR_LIB_IN",   "GAL_LIB_OUT",
                            "QSO_LIB_OUT", "STAR_LIB_OUT",  "LIB_ASCII",
                            "EM_LINES",    "EM_DISPERSION", "ADD_DUSTEM",
                            "VERBOSE"};
  // Number of keywords
  int nb_ref_key = (int)(sizeof(list_keywords) / sizeof(list_keywords[0]));

  // Find the reference keywords within the command line + config file -> store
  // them into a vector
  keymap key_analysed = analyse_keywords(argc, argv, list_keywords, nb_ref_key);

  // keyword to add the LDUST component to the stellar component (e.g. in BC03)
  bool add_dust = key_analysed["ADD_DUSTEM"].split_bool("NO", 1)[0];

  // Define a pointer of the basis class "Mag" which encompasses all the
  // elements to create the library
  Mag *Magnitude;
  object_type object = SED::string_to_object(key_analysed["t"].value);
  switch (object) {
    case GAL:
      Magnitude = new GalMag(key_analysed);
      break;
    case QSO:
      Magnitude = new QSOMag(key_analysed);
      break;
    case STAR:
      Magnitude = new StarMag(key_analysed);
      break;
  }

  /*
  OPEN FILES
  */

  Magnitude->open_files();

  /*
  INFO PARAMETERS ON SCREEN AND DOC
  */

  Magnitude->print_info();

  /*
  READ DUST EXTINCTION LAWS
  */

  Magnitude->read_ext();

  /*
  DEFINE THE REDSHIFT GRID
  */

  Magnitude->def_zgrid();

  /*
  READ B12 TEMPLATES TO ADD DUST EMISSION TO BC03
  */

  if (add_dust) {
    Magnitude->read_B12();
  }

  /*
  READ SED, APPLY EXTINCTION AND IGM OPACITY
  */

  Magnitude->read_SED();

  /*
  WRITE DOCUMENTATION
  */
  Magnitude->write_doc();

  /*
  DONE
  */

  delete Magnitude;

  // display the time needed to run the code
#ifdef _OPENMP
  double end = omp_get_wtime();
  double diff = end - start;
#else
  clock_t t2 = clock();
  float diff(((float)t2 - (float)t1) / CLOCKS_PER_SEC);
#endif

  cout << endl << "Times to run the code in sec: " << diff << endl;

  return 0;
}

File: src/lephare/mag_gal.py
from contextlib import suppress

from ._lephare import GalMag, QSOMag, StarMag, keyword
from .runner import Runner

__all__ = [
    "MagGal",
]

config_keys = {
    "typ": "define what kind of objects these SED belong to : GAL, QSO, or STAR",
    "verbose": "increase onscreen verbosity",
    "COSMOLOGY": "fiducial cosmology used for absolute magnitudes evaluations",
    "FILTER_FILE": "filter file provided by filter script or the Filter class",
    "MAGTYPE": "AB or VEGA system",
    "EXTINC_LAW": "list of extinction files used for extinction correction\
    (if relative, in $LEPHAREDIR/ext/)",
    "MOD_EXTINC": "list of SED template ranges for which each extinction law is going to be applied",
    "EB_V": "E(B-V) grid of values to be applied to the SEDs (comma separated)",
    "Z_STEP": "dz, zmin, zmax defining the redshift grid",
    "GAL_LIB_IN": "input SED library file, as output by sedtolib script or SEDtolib class\
    (in $LEPHAREWORK/lib_bin/)",
    "GAL_LIB_OUT": "output file of the synthetic magnitudes (in $LEPHAREWORK/lib_mag/)",
    "QSO_LIB_IN": "same for QSO/AGN templates",
    "QSO_LIB_OUT": "same for QSO/AGN templates",
    "STAR_LIB_IN": "same for STAR templates",
    "STAR_LIB_OUT": "same for STAR templates",
    "LIB_ASCII": "if set to YES, also provide the output in ascii",
    "EM_LINES": "[NO/EMP_UV/EMP_SFR/PHYS] choice of prescription for emission line computation",
    "EM_DISPERSION": "rescaling values for the emission lines",
    "ADD_DUSTEM": "add the dust emission in templates when missing",
}


class MagGal(Runner):
    """The specific arguments to the MagGal class are

    typ:
           define what kind of objects these SED belong to : GAL, QSO, or STAR
    verbose:
           increase onscreen verbosity
    COSMOLOGY:
           fiducial h0, Omega_m0, and LambdaO used to define a flat LCDM cosmology
           used for absolute magnitudes evaluations
    FILTER_FILE:
           filter file provided by filter script or the Filter class
    MAGTYPE:
           AB or VEGA system
    EXTINC_LAW:
           list of extinction files used for extinction correction
           (if relative, in $LEPHAREDIR/ext/)
    MOD_EXTINC:
           list of SED template ranges for which each extinction law is going to be applied
    EB_V:
           E(B-V) grid of values to be applied to the SEDs (comma separated)
    Z_STEP:
           dz, zmin, zmax defining the redshift grid
    GAL_LIB_IN:
           input SED library file, as output by sedtolib script or SEDtolib class
           (in $LEPHAREWORK/lib_bin/)
    GAL_LIB_OUT:
           output file of the synthetic magnitudes (in $LEPHAREWORK/lib_mag/)
    QSO_LIB_IN, QSO_LIB_OUT, STAR_LIB_IN, STAR_LIB_OUT
           same for STAR templates
    LIB_ASCII:
           if set to YES, also provide the output in ascii
    EM_LINES:
           [NO/EMP_UV/EMP_SFR/PHYS] choice of prescription for emission line computation
    EM_DISPERSION:
           possible rescaling values for the emission lines
    ADD_DUSTEM:
           add the dust emission in templates when missing
    VERBOSE:
           add verbosity
    """

    def update_help(self):
        """Add the specific MagGal help"""
        doc = "Build the LePHARE internal representation of the set of SED templates to be used"
        with suppress(Exception):
            self.parser.usage = "Build the LePHARE synthetic magnitudes"
        self.__doc__ = doc + "\n"  # + inspect.getdoc(MagGal)

    def __init__(self, config_file=None, config_keymap=None, **kwargs):
        super().__init__(config_keys, config_file, config_keymap, **kwargs)

    def run(self, **kwargs):
        """Compute the model magnitudes across the redshift grid.

        Returns
        -------
        None
        """
        super().run(**kwargs)

        # needed to comply with the C++ interface
        # Define the type (Galaxy, QSO, Stars)
        self.keymap["t"] = keyword("t", self.typ)
        # Parameter file
        self.keymap["c"] = keyword("c", self.config)

        if self.typ[0] == "G":
            mag = GalMag(self.keymap)
        elif self.typ[0] == "Q":
            mag = QSOMag(self.keymap)
        elif self.typ[0] == "S":
            mag = StarMag(self.keymap)
        else:
            raise KeyError("-t arg must start with G/g Q/q or S/s for Galaxy QSO and Star respectively.")
        mag.open_files()
        mag.print_info()
        # Read dust extinction laws
        mag.read_ext()
        # Define the redshift grid
        mag.def_zgrid()
        # Read B12 templates to add dust emission to BC03
        add_dust = self.keymap["ADD_DUSTEM"].split_bool("NO", 1)
        if add_dust:
            mag.read_B12()
        # Read sed, apply extinction and IGM opacity
        mag.read_SED()
        mag.write_doc()
        # we need to call the close method here because run can
        # be called within a python session that stays alive afterwards
        mag.close_files()

        self.Mag = mag
        return


def main():  # pragma no cover
    runner = MagGal()
    runner.run()
    runner.end()


if __name__ == "__main__":  # pragma no cover
    main()
